testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs

```diff
@@ -0,0 +1,517 @@
+        stackerdb_conn: &StackerDBs,
+        miners_db_config: StackerDBConfig,
+        config: &Config,
+    ) -> Result<Self, ChainstateError> {
+        let Some(ref reward_set_signers) = reward_set.signers else {
+            error!("Could not initialize WSTS coordinator for reward set without signer");
+            return Err(ChainstateError::NoRegisteredSigners(0));
+        };
+
+        let Some(receiver) = STACKER_DB_CHANNEL
+            .receiver
+            .lock()
+            .expect("FATAL: StackerDBChannel lock is poisoned")
+            .take()
+        else {
+            error!("Could not obtain handle for the StackerDBChannel");
+            return Err(ChainstateError::ChannelClosed(
+                "WSTS coordinator requires a handle to the StackerDBChannel".into(),
+            ));
+        };
+
+        let NakamotoSigningParams {
+            num_signers,
+            num_keys,
+            threshold,
+            signer_key_ids,
+            signer_public_keys,
+            wsts_public_keys,
+        } = NakamotoSigningParams::from(reward_set_signers.as_slice());
+        debug!(
+            "Initializing miner/coordinator";
+            "num_signers" => num_signers,
+            "num_keys" => num_keys,
+            "threshold" => threshold,
+            "signer_key_ids" => ?signer_key_ids,
+            "signer_public_keys" => ?signer_public_keys,
+            "wsts_public_keys" => ?wsts_public_keys,
+        );
+        let coord_config = CoordinatorConfig {
+            num_signers,
+            num_keys,
+            threshold,
+            signer_key_ids,
+            signer_public_keys,
+            dkg_threshold: threshold,
+            message_private_key: message_key.clone(),
+            ..Default::default()
+        };
+
+        let mut coordinator: FireCoordinator<Aggregator> = FireCoordinator::new(coord_config);
+        let party_polynomials = get_signer_commitments(
+            is_mainnet,
+            reward_set_signers.as_slice(),
+            stackerdb_conn,
+            reward_cycle,
+            &aggregate_public_key,
+        )?;
+        if let Err(e) = coordinator
+            .set_key_and_party_polynomials(aggregate_public_key.clone(), party_polynomials)
+        {
+            warn!("Failed to set a valid set of party polynomials"; "error" => %e);
+        };
+
+        Ok(Self {
+            coordinator,
+            message_key,
+            receiver: Some(receiver),
+            wsts_public_keys,
+            is_mainnet,
+            miners_db_config,
+            signing_round_timeout: config.miner.wait_on_signers.clone(),
+        })
+    }
+
+    fn get_sign_id(burn_block_height: u64, burnchain: &Burnchain) -> u64 {
+        burnchain
+            .pox_constants
+            .reward_cycle_index(burnchain.first_block_height, burn_block_height)
+            .expect("FATAL: tried to initialize WSTS coordinator before first burn block height")
+    }
+
+    fn send_signers_message(
```

Comment:
Is this meant to broadcast the message to other nodes?  Because, this will be pretty slow -- it requires waiting for the StackerDB state machine to synchronize DB inventories with its neighbors, notice that it has a newer slot version than everybody else, and then broadcast the new chunk to them.  Instead, this function could directly send the StackerDB chunk to connected peers that replicate this chunk.

Reply:
This might take a bit of code, but it's not much. You'd want to give `SignerCoordinator` a `NetworkHandle` instance connected to the `PeerNetwork` instance (we already have one in the `Relayer`).  From there, you can queue messages to broadcast on the p2p network via `::send_message()`.  We'd want to extend the logic to specifically accommodate StackerDB messages so we only send the message to a peer that replicates the same StackerDB, but that wouldn't be too bad.  I can add this on top of this PR (or as a follow-on one) if you don't want to add it here.

Reply:
That sounds like it definitely needs to be a separate PR.

Reply:
Switched to using the RPC interface, which avoids this issue for now.